### PR TITLE
Add new lint: `strlen_on_c_strings`

### LINKED ISSUE #7145: lint on `libc::strlen(some_cstr.as_ptr())`
### What it does

This lint would flag the use of `libc::strlen` on a `CString` or `CStr` value, and suggest calling `to_bytes().len()` instead.

### Categories (optional)

- Kind: `clippy::complexity`

*What is the advantage of the recommended code over the original code*

This avoids calling an unsafe `libc` function. Currently, it also avoids calculating the length; in the future, this may become equivalent to calling `strlen` at runtime, but it'll still avoid the need for an unsafe `libc` call.

### Drawbacks

None.

### Example

```rust
libc::strlen(name.as_ptr())
```

Could be written as:

```rust
name.to_bytes().len()
```

### PR DESCRIPTION
**Pull request by mgacek8**
~~This is WIP, linting in case of `CString` has been added, but for `CStr`, its diagnostic item needs to be available for clippy. 
[PR that adds diagnostic item for CStr on rust repo](https://github.com/rust-lang/rust/pull/85439).~~
Ready for the review. Please take a look. 
fixes #7145
changelog: Add new lint: `strlen_on_c_strings`, that lints on `libc::strlen(some_cstring.as_ptr())`

### CODE REVIEW

File: clippy_lints/src/strlen_on_c_strings.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+declare_lint_pass!(StrlenOnCStrings => [STRLEN_ON_C_STRINGS]);
+
+impl LateLintPass<'tcx> for StrlenOnCStrings {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if in_macro(expr.span) {
+            return;
+        }
+
+        if_chain! {
+            if let hir::ExprKind::Call(func, args) = expr.kind;
```
Comment (camsteffen) [inline clippy_lints/src/strlen_on_c_strings.rs:47]: drive-by nit
```suggestion
            if let hir::ExprKind::Call(func, [recv]) = expr.kind;
```

**Comment by giraffate** [comment]
ping from triage @mgacek8. https://github.com/rust-lang/rust/pull/85439 has been merged, so is this ready for review?

**Comment by mgacek8** [comment]
> 
> 
> ping from triage @mgacek8. [rust-lang/rust#85439](https://github.com/rust-lang/rust/pull/85439) has been merged, so is this ready for review?

Not yet, I resolved conflicts, but I need to investigate why `is_type_diagnostic_item(cx, ty, sym::CStr)` doesn't return true when type is `CStr` and fix it.

**Comment by mgacek8** [comment]
It is ready for the review, please take a look. Can we change the label ?

File: clippy_lints/src/strlen_on_c_strings.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                let val_name = snippet_with_macro_callsite(cx, cstring.span, "..");
+                let sugg = if is_type_diagnostic_item(cx, ty, sym::cstring_type){
+                    format!("{}.as_bytes().len()", val_name)
+                }
+                else if is_type_ref_to_diagnostic_item(cx, ty, sym::CStr){
+                    format!("{}.to_bytes().len()", val_name)
+                }
+                else {
+                    return;
+                };
```
Comment (giraffate) [inline clippy_lints/src/strlen_on_c_strings.rs:68]: Doesn't rustfmt work properly?
```suggestion
                let sugg = if is_type_diagnostic_item(cx, ty, sym::cstring_type){
                    format!("{}.as_bytes().len()", val_name)
                } else if is_type_ref_to_diagnostic_item(cx, ty, sym::CStr){
                    format!("{}.to_bytes().len()", val_name)
                } else {
                    return;
                };
```

File: clippy_lints/src/strlen_on_c_strings.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        if in_macro(expr.span) {
+            return;
+        }
+
+        if_chain! {
+            if let hir::ExprKind::Call(func, [recv]) = expr.kind;
+            if let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = func.kind;
+
+            if (&paths::LIBC_STRLEN).iter().map(|x| Symbol::intern(x)).eq(
+                path.segments.iter().map(|seg| seg.ident.name));
```
Comment (giraffate) [inline clippy_lints/src/strlen_on_c_strings.rs:51]: Can `match_def_path` util method be used here?

File: clippy_lints/src/strlen_on_c_strings.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_ref_to_diagnostic_item};
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::{sym, Symbol};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for usage of `libc::strlen` on a `CString` or `CStr` value, and suggest calling `as_bytes().len()` or `to_bytes().len()` respectively instead.
```
Comment (giraffate) [inline clippy_lints/src/strlen_on_c_strings.rs:14]: nits
```suggestion
    /// **What it does:** Checks for usage of `libc::strlen` on a `CString` or `CStr` value,
    /// and suggest calling `as_bytes().len()` or `to_bytes().len()` respectively instead.
```

File: clippy_lints/src/strlen_on_c_strings.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    ///
+    /// **Why is this bad?** This avoids calling an unsafe `libc` function. Currently, it also avoids calculating the length.
```
Comment (giraffate) [inline clippy_lints/src/strlen_on_c_strings.rs:16]: nits
```suggestion
    /// **Why is this bad?** This avoids calling an unsafe `libc` function.
    /// Currently, it also avoids calculating the length.
```

**Review by giraffate — COMMENTED** [review]
Overall looks good, thanks! I made some comments.

File: clippy_lints/src/strlen_on_c_strings.rs (hunk 3 above)
Comment (mgacek8) [reply clippy_lints/src/strlen_on_c_strings.rs:51]: There was a problem that `let path = cx.get_def_path(did);` returns the following path (on Windows):
```
 [clippy_utils\src\lib.rs:1265] &path = [
     "libc",
     "windows",
     "",
     "strlen",
 ]
```

File: clippy_lints/src/strlen_on_c_strings.rs (hunk 2 above)
Comment (mgacek8) [reply clippy_lints/src/strlen_on_c_strings.rs:68]: Seems like no inside of a macro

File: clippy_lints/src/strlen_on_c_strings.rs (hunk 3 above)
Comment (giraffate) [reply clippy_lints/src/strlen_on_c_strings.rs:53]: Oh, you're right.

File: tests/ui/strlen_on_c_strings.stderr [hunk 6]
```diff
@@ -0,0 +1,16 @@
+error: using `libc::strlen` on a `CString` or `CStr` value
+  --> $DIR/strlen_on_c_strings.rs:11:24
+   |
+LL |     let len = unsafe { libc::strlen(cstring.as_ptr()) };
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `cstring.as_bytes().len()`
```
Comment (giraffate) [inline tests/ui/strlen_on_c_strings.stderr:5]: The suggested code don't need unsafe block. Can unsafe block be checked? Or this comments needs to be improved to include removing unsafe block.

Comment (mgacek8) [reply tests/ui/strlen_on_c_strings.stderr:5]: Good catch, yes, unsafe block is redundant here. It can be removed when it's the only one unsafe operation inside of it.
What about for now to lint like this: "help: try this: `cstring.as_bytes().len()`, you might also need to get rid of `unsafe` block in some cases"?

Comment (giraffate) [reply tests/ui/strlen_on_c_strings.stderr:5]: Yes, it looks good as a first step! That unsafe block might be removed is one of purposes of this lint.

If unsafe block has only `libc::strlen(cstr.as_ptr())`, it would be better to remove unsafe block in suggestion. But I think it can be fixed with another follow-up PR.

File: tests/ui/strlen_on_c_strings.stderr [hunk 7]
```diff
@@ -0,0 +1,16 @@
+error: using `libc::strlen` on a `CString` or `CStr` value
+  --> $DIR/strlen_on_c_strings.rs:11:24
+   |
+LL |     let len = unsafe { libc::strlen(cstring.as_ptr()) };
+   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `cstring.as_bytes().len(), you might also need to get rid of `unsafe` block in some cases`
```
Comment (giraffate) [inline tests/ui/strlen_on_c_strings.stderr:5]: Can this comment be fixed like this?
```suggestion
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `cstring.as_bytes().len()`, you might also need to get rid of `unsafe` block in some cases
```

Comment (mgacek8) [reply tests/ui/strlen_on_c_strings.stderr:5]: Seems like `span_lint_and_sugg` doesn't allow that. 
What about: "help: try this (you might also need to get rid of `unsafe` block in some cases): `cstring.as_bytes().len()`" ?

Comment (giraffate) [reply tests/ui/strlen_on_c_strings.stderr:5]: Ok, it looks good!

**Review by giraffate — APPROVED** [review]

**Comment by giraffate** [comment]
@bors r+

Thanks!

File: tests/ui/strlen_on_c_strings.stderr (hunk 6 above)
Comment (mgacek8) [reply tests/ui/strlen_on_c_strings.stderr]: Created the issue for that: https://github.com/rust-lang/rust-clippy/issues/7436
